### DISCUSSION TITLE
more than 200 LHS where clauses make knowledge building really slow

### DISCUSSION THREAD

**Discussion by rohan3usturge** [evrete/evrete#65] [Q&A] [answered]
knowledge
.newRule()
.forEach("$a", SomeVeryLargeClass.class "$b",Result.class)
.where(...more than 200 clauses)
.execute(ctx -> {
  Result result = ctx.get("$b");
  result.updateSuccess(true);
}

execute is really slow

Can you help how can I optimize this ?

**Comment by andbi** [comment]
I will look into it. As of now, the engine launches internal Java compiler for every literal condition it encounters and compiles them into java byte code one by one. Bulk compilation could be a better option here.

However, it's important to remember that building a Knowledge instance isn't supposed to be fast. The most common approach involves building Knowledge instances once and spawning knowledge sessions off of them.

**Reply by andbi** [reply, marked as answer]
@rohan3usturge could you please check if the compilation time is any better in 3.0.04? In your particular case, I expect it to be faster by orders of magnitude.

The new version bulk compiles all the conditions within a single rule, the next version will compile the whole ruleset in a single operation.


And thank you for reporting the issue, that was a serious one.

regards, @andbi
